wfe/resource.go

```diff
@@ -59,47 +60,68 @@ func (r *resource) Run(c eval.Context, input eval.OrderedMap) eval.OrderedMap {
 		}
 	}
 
-	var result eval.Value
+	var result eval.PuppetObject
 	hn := handlerDef.Identifier().Name()
 	switch op {
 	case wfapi.Read:
 		if extId == nil {
 			return eval.EMPTY_MAP
 		}
-		result = handler.Invoke(c, hn, `read`, extId).(eval.OrderedMap)
+		result = eval.AssertInstance(handlerDef.Label, r.typ, handler.Invoke(c, hn, `read`, extId)).(eval.PuppetObject)
```

Comment:
assume if handler.Invoke results in a panic in the plugin, this will not obscure that error?

Reply:
No, that panic will never come back to the caller of Invoke. It must be handled by a recover() higher up.